include/llvm-dialects/Dialect/ContextExtension.h

```diff
@@ -0,0 +1,112 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "llvm/ADT/ArrayRef.h"
+
+namespace llvm {
+class LLVMContext;
+} // namespace llvm
+
+namespace llvm_dialects {
+
+class DialectContext;
+
+namespace detail {
+
+class ContextExtensionKey {
+  using CreateFn = void *(llvm::LLVMContext &);
+  using DestroyFn = void(llvm::LLVMContext &, void *);
+
+  unsigned m_index;
+  CreateFn *m_create = nullptr;
+  DestroyFn *m_destroy = nullptr;
+
+public:
+  ContextExtensionKey(CreateFn *createFn, DestroyFn *destroyFn);
+  ~ContextExtensionKey();
+
+  unsigned getIndex() const { return m_index; }
+  void *create(llvm::LLVMContext &context);
+  void destroy(llvm::LLVMContext &context, void *extension);
+
+  void *get(llvm::LLVMContext &context);
+  void *get(DialectContext &context);
+
+  static llvm::ArrayRef<ContextExtensionKey *> getKeys();
+};
+
+} // namespace detail
+
+/// CRTP base class for extended data that can be attached to LLVMContexts.
+///
+/// In fact, extended data is currently attached to the DialectContext, but
+/// conceptually this feature is independent of dialect support.
+///
+/// Extended data types must derive from this template base using the following
+/// pattern:
+///
+///   class MyExtension : public ContextExtensionImpl<MyExtension> {
+///   public:
+///     MyExtension(llvm::LLVMContext &context);
+///
+///     static Key theKey; // instantiate this in a source file
+///   };
+///
+/// Extension data can be retrieved using the static method call
+/// MyExtension::get(llvmContext).
+///
+/// The extension data is allocated and constructed the first time it is
+/// requested. The destructor is called when the DialectContext is destroyed.
+///
+/// Note that all context extension types must be present (their keys having
+/// registered themselves) _before_ creating the DialectContext with which they
+/// are going to be used (this restriction only really matters when dynamic
+/// linking is used: it is not possible to dynamically load an extension after
+/// a context is created, and then use the extension with that context).
+template <typename ContextExtensionT> class ContextExtensionImpl {
+public:
+  class Key : public detail::ContextExtensionKey {
+    static void *createImpl(llvm::LLVMContext &context) {
+      return static_cast<void *>(new ContextExtensionT(context));
+    }
+
+    static void destroyImpl(llvm::LLVMContext &context, void *untyped) {
+      ContextExtensionT *extension = static_cast<ContextExtensionT *>(untyped);
+      delete extension;
+    }
```

Comment:
Is the manual handling of function pointers really necessary? 

What if we introduced a common, non-templated base class `ContextExtensionImplBase` of all ContextExtensionImpl that has a virtual destructor,  and replaced the `void*` by  `std::unique_ptr<ContextExtensionImplBase>`? 

For creation, we could use a virtual function in `Key`.

Reply:
We could also use a virtual function for `destroy` even without a virtual destructor, but also getting rid of the manual memory management and `void*` casts may be even nicer?

Reply:
Yeah, turns out that this can work out nicely after all.

Reply:
Nice!